File: ios/Classes/TotalProcessingPlugin.swift
import Flutter
import SafariServices

public class TotalProcessingPlugin: UIViewController, FlutterPlugin, FlutterStreamHandler, SFSafariViewControllerDelegate {
    static let handleCheckoutResultEvent = "handleCheckoutResult"
    private var handleCheckoutResultSink: FlutterEventSink?
    
    static let customUIResultEvent = "customUIResult"
    private var customUIResultSink: FlutterEventSink?
    private var safariVC: SFSafariViewController?
    
    let provider = OPPPaymentProvider(mode: OPPProviderMode.live)
    let checkoutSettings = OPPCheckoutSettings()
    
    let threeDSChallengeViewController = ThreeDSChallengeViewController()
    
    override public func viewDidLoad() {
        super.viewDidLoad()
        self.provider.threeDSEventListener = self.threeDSChallengeViewController
    }
    
    func presenterURL(url: URL) {
        self.safariVC = SFSafariViewController(url: url)
        self.safariVC?.delegate = self
        self.present(self.safariVC!, animated: true, completion: nil)
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        switch arguments as? String {
        case TotalProcessingPlugin.handleCheckoutResultEvent:
            self.handleCheckoutResultSink = events
        case TotalProcessingPlugin.customUIResultEvent:
            self.customUIResultSink = events
        default:
            break
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "total_processing", binaryMessenger: registrar.messenger())
        let instance = TotalProcessingPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        let handleCheckoutResultChannel = FlutterEventChannel(name: handleCheckoutResultEvent, binaryMessenger: registrar.messenger())
        handleCheckoutResultChannel.setStreamHandler(instance)
        
        let customUIResultChannel = FlutterEventChannel(name: customUIResultEvent, binaryMessenger: registrar.messenger())
        customUIResultChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startCheckout":
            if let args = call.arguments as? [String: Any] {
                let checkoutID = args["checkoutId"] as! String
            
                let paymentBrands = args["paymentBrands"] as! [String]
                self.checkoutSettings.paymentBrands = paymentBrands
            
                let shopperResultURL = args["shopperResultURL"] as! String
                self.checkoutSettings.shopperResultURL = shopperResultURL
            
                self.startCheckout(checkoutID: checkoutID)
            } else {
                result(FlutterError(code: "InvalidArgumentt", message: "Invalid argument", details: nil))
            }
        case "customUIPay":
            if let args = call.arguments as? [String: Any] {
                let checkoutID = args["checkoutId"] as! String
                let cardHolder = args["cardHolder"] as! String
                let cardNumber = args["cardNumber"] as! String
                if !OPPCardPaymentParams.isNumberValid(cardNumber, luhnCheck: true) {
                    var item: [String: Any?] = [:]
                    item["isErrored"] = true
                    var paymentError: [String: Any?] = [:]
                    paymentError["errorCode"] = "cardNumber"
                    paymentError["errorInfo"] = "Invalid card number"
                    paymentError["errorMessage"] = "Invalid card number"
                    item["paymentError"] = paymentError
                    self.customUIResultSink?(item)
                } else {
                    let expiryMonth = args["expiryMonth"] as! String
                    let expiryYear = args["expiryYear"] as! String
                    let cvc = args["cvc"] as! String
                    let cardBrand = args["cardBrand"] as! String
                
                    let shopperResultURL = args["shopperResultUrl"] as! String
           
                    guard let transactionX = self.createTransaction(checkoutID: checkoutID, cardHolder: cardHolder, cardNumber: cardNumber, expiryMonth: expiryMonth, expiryYear: expiryYear, cvc: cvc, cardBrand: cardBrand, shopperResultURL: shopperResultURL) else {
                        var item: [String: Any?] = [:]
                        item["isCanceled"] = true
                        self.customUIResultSink?(item)
                        return
                    }
                    
                    self.provider.submitTransaction(transactionX, completionHandler: { transaction, error in
                        DispatchQueue.main.async {
                            var item: [String: Any?] = [:]
                            if error != nil {
                                item["isErrored"] = true
                                var paymentError: [String: Any?] = [:]
                                paymentError["errorCode"] = "401"
                                paymentError["errorInfo"] = "error"
                                paymentError["errorMessage"] = error?.localizedDescription
                                item["paymentError"] = paymentError
                                self.customUIResultSink?(item)
                                return
                            }
                            
                            var trans: [String: Any?] = [:]
                            if transaction.type == .synchronous {
                                trans["transactionType"] = "synchronous"
                            } else if transaction.type == .asynchronous {
                                trans["transactionType"] = "asynchronous"
                            }
                            var paymentParams: [String: Any?] = [:]
                            paymentParams["checkoutId"] = checkoutID
                            trans["paymentParams"] = paymentParams
                            trans["brandSpecificInfo"] = transaction.brandSpecificInfo
                            trans["redirectUrl"] = transaction.redirectURL?.absoluteString
                            var threeDS2Info: [String: Any?] = [:]
                            threeDS2Info["authResponse"] = transaction.threeDS2Info?.authResponse
                            threeDS2Info["authStatus"] = transaction.threeDS2Info?.authStatus.rawValue
                            threeDS2Info["callbackURL"] = transaction.threeDS2Info?.callbackURL
                            threeDS2Info["cardHolderInfo"] = transaction.threeDS2Info?.cardHolderInfo
                            threeDS2Info["challengeCompletionCallbackUrl"] = transaction.threeDS2Info?.challengeCompletionCallbackUrl
                            threeDS2Info["protocolVersion"] = transaction.threeDS2Info?.protocolVersion
                            threeDS2Info["schemeConfig"] = transaction.threeDS2Info?.schemeConfig?.description
                            threeDS2Info["threeDSFlow"] = transaction.threeDS2Info?.threeDSFlow.rawValue
                            trans["threeDS2Info"] = threeDS2Info
                            trans["threeDS2MethodRedirectUrl"] = transaction.threeDS2MethodRedirectURL?.absoluteString
                            trans["yooKassaInfo"] = transaction.yooKassaInfo
                            item["transaction"] = trans
                            self.customUIResultSink?(item)
                        }
                    })
                }

            } else {
                result(FlutterError(code: "InvalidArgumentt", message: "Invalid argument", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func createTransaction(checkoutID: String, cardHolder: String, cardNumber: String, expiryMonth: String, expiryYear: String, cvc: String, cardBrand: String, shopperResultURL: String) -> OPPTransaction? {
        var item: [String: Any?] = [:]
        do {
            let params = try OPPCardPaymentParams(checkoutID: checkoutID, paymentBrand: cardBrand, holder: cardHolder, number: cardNumber, expiryMonth: expiryMonth, expiryYear: "20\(expiryYear)", cvv: cvc)
            params.shopperResultURL = shopperResultURL
            return OPPTransaction(paymentParams: params)
        } catch let error as NSError {
            item["isErrored"] = true
            var paymentError: [String: Any?] = [:]
            paymentError["errorCode"] = "\(error.code)"
            paymentError["errorInfo"] = "\(error.domain)"
            paymentError["errorMessage"] = error.localizedDescription
            item["paymentError"] = paymentError
            self.customUIResultSink?(item)
            return nil
        }
    }
    
    private func startCheckout(checkoutID: String) {
        let config = OPPThreeDSConfig()
        self.checkoutSettings.threeDSConfig = config
        let checkoutProvider = OPPCheckoutProvider(paymentProvider: provider, checkoutID: checkoutID, settings: checkoutSettings)
        var item: [String: Any?] = [:]
        // Since version 2.13.0
        checkoutProvider?.presentCheckout(forSubmittingTransactionCompletionHandler: { transaction, error in
            guard let transaction = transaction else {
                item["isErrored"] = true
                var paymentError: [String: Any?] = [:]
                paymentError["errorCode"] = "401"
                paymentError["errorInfo"] = "error"
                paymentError["errorMessage"] = error?.localizedDescription
                item["paymentError"] = paymentError
                self.handleCheckoutResultSink?(item)
                return
            }
            
            var trans: [String: Any?] = [:]
            if transaction.type == .synchronous {
                trans["transactionType"] = "synchronous"
            } else if transaction.type == .asynchronous {
                trans["transactionType"] = "asynchronous"
            }
            var paymentParams: [String: Any?] = [:]
            paymentParams["checkoutId"] = checkoutID
            trans["paymentParams"] = paymentParams
            trans["brandSpecificInfo"] = transaction.brandSpecificInfo
            trans["redirectUrl"] = transaction.redirectURL?.absoluteString
            var threeDS2Info: [String: Any?] = [:]
            threeDS2Info["authResponse"] = transaction.threeDS2Info?.authResponse
            threeDS2Info["authStatus"] = transaction.threeDS2Info?.authStatus.rawValue
            threeDS2Info["callbackURL"] = transaction.threeDS2Info?.callbackURL
            threeDS2Info["cardHolderInfo"] = transaction.threeDS2Info?.cardHolderInfo
            threeDS2Info["challengeCompletionCallbackUrl"] = transaction.threeDS2Info?.challengeCompletionCallbackUrl
            threeDS2Info["protocolVersion"] = transaction.threeDS2Info?.protocolVersion
            threeDS2Info["schemeConfig"] = transaction.threeDS2Info?.schemeConfig?.description
            threeDS2Info["threeDSFlow"] = transaction.threeDS2Info?.threeDSFlow.rawValue
            trans["threeDS2Info"] = threeDS2Info
            trans["threeDS2MethodRedirectUrl"] = transaction.threeDS2MethodRedirectURL?.absoluteString
            trans["yooKassaInfo"] = transaction.yooKassaInfo
            item["transaction"] = trans
            self.handleCheckoutResultSink?(item)
        }, cancelHandler: {
            // Executed if the shopper closes the payment page prematurely
            item["isCanceled"] = true
            self.handleCheckoutResultSink?(item)
        })
    }
}

class ThreeDSChallengeViewController: UINavigationController, OPPThreeDSEventListener {
    public func onThreeDSChallengeRequired(completion: @escaping (UINavigationController) -> Void) {
        completion(self)
    }
    
    public func onThreeDSConfigRequired(completion: @escaping (OPPThreeDSConfig) -> Void) {
        let config = OPPThreeDSConfig()
        completion(config)
    }
}

File: android/src/main/kotlin/ae/vire/total_processing/TotalProcessingPlugin.kt
@file:Suppress("NAME_SHADOWING")

package ae.vire.total_processing

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Parcelable
import android.util.Log
import com.oppwa.mobile.connect.checkout.dialog.CheckoutActivity
import com.oppwa.mobile.connect.checkout.meta.CheckoutActivityResult
import com.oppwa.mobile.connect.checkout.meta.CheckoutSettings
import com.oppwa.mobile.connect.exception.PaymentError
import com.oppwa.mobile.connect.exception.PaymentException
import com.oppwa.mobile.connect.payment.card.CardPaymentParams
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import com.oppwa.mobile.connect.provider.*


/** TotalProcessingPlugin */
class TotalProcessingPlugin: FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener{
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var activity: Activity

  private val providerMode= Connect.ProviderMode.LIVE

  private var paymentProvider: OppPaymentProvider? = null


  private val checkoutRequestCode = 123

  private var handleCheckoutResultEvent: EventChannel? = null
  private var handleCheckoutResultSink : EventChannel.EventSink? = null
  private val handleCheckoutResultHandler = object : EventChannel.StreamHandler {
    override fun onListen(arg: Any?, eventSink: EventChannel.EventSink?) {
      handleCheckoutResultSink = eventSink
    }
    override fun onCancel(o: Any?) {}
  }

  private var customUIResultEvent: EventChannel? = null
  private var customUIResultSink : EventChannel.EventSink? = null
  private val customUIResultHandler = object : EventChannel.StreamHandler {
    override fun onListen(arg: Any?, eventSink: EventChannel.EventSink?) {
      customUIResultSink = eventSink
    }
    override fun onCancel(o: Any?) {}
  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "total_processing")
    channel.setMethodCallHandler(this)
    handleCheckoutResultEvent = EventChannel(flutterPluginBinding.binaryMessenger, "handleCheckoutResult")
    handleCheckoutResultEvent!!.setStreamHandler(handleCheckoutResultHandler)

    customUIResultEvent = EventChannel(flutterPluginBinding.binaryMessenger, "customUIResult")
    customUIResultEvent!!.setStreamHandler(customUIResultHandler)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when(call.method) {
      "startCheckout" -> {
        val checkoutID: String = call.argument<String>("checkoutId")!!
        val paymentBrands: List<String> = call.argument<List<String>>("paymentBrands")!!
          startCheckout(checkoutID, paymentBrands)
      }
      "customUIPay" -> {
        val checkoutID: String = call.argument<String>("checkoutId")!!
        val cardHolder: String = call.argument<String>("cardHolder")!!
        val cardNumber: String = call.argument<String>("cardNumber")!!
        if (!CardPaymentParams.isNumberValid(cardNumber,true)) {
          val item: MutableMap<String, Any?> = HashMap()
          item["isErrored"] = true
          val paymentError: MutableMap<String, Any?> = HashMap()
          paymentError["errorCode"] = "cardNumber"
          paymentError["errorInfo"] = "Invalid card number"
          paymentError["errorMessage"] = "Invalid card number"
          item["paymentError"] = paymentError
          handleCheckoutResultSink?.success(item)
          return
        }
        val expiryMonth: String = call.argument<String>("expiryMonth")!!
        val expiryYear: String = call.argument<String>("expiryYear")!!
        val cvc: String = call.argument<String>("cvc")!!
        val cardBrand: String = call.argument<String>("cardBrand")!!
        val shopperResultUrl: String = call.argument<String>("shopperResultUrl")!!
        pay(checkoutID,cardHolder,cardNumber,expiryMonth,expiryYear,cvc,cardBrand, shopperResultUrl)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  @SuppressLint("SuspiciousIndentation")
  private fun startCheckout(checkoutID: String, paymentBrands: List<String>) {
    // since mSDK version 6.0.0 the shopper result URL is not required
    val checkoutSettings = CheckoutSettings(checkoutID, paymentBrands.toSet(), providerMode)
    val var3: Intent = Intent(
      activity,
      CheckoutActivity::class.java
    ).putExtra("com.oppwa.mobile.connect.checkout.dialog.EXTRA_CHECKOUT_SETTINGS", checkoutSettings)
      if (checkoutSettings.componentName != null) {
        var3.putExtra(
          "com.oppwa.mobile.connect.checkout.dialog.EXTRA_CHECKOUT_RECEIVER",
          checkoutSettings.componentName
        )
      }
      if (checkoutSettings.paymentButtonBrand != null) {
        var3.putExtra(
          "com.oppwa.mobile.connect.checkout.dialog.EXTRA_CHECKOUT_PAYMENT_BUTTON_METHOD",
          checkoutSettings.paymentButtonBrand
        )
      }
    Log.i("startActivity", "startActivity")
    activity.startActivityForResult(var3,checkoutRequestCode)
  }


  private fun handleCheckoutResult(result: CheckoutActivityResult) {
//    if (result.isErrored) {
//      Log.i("handleCheckoutResult", "errorInfo :${result.paymentError?.errorInfo}")
//      Log.i("handleCheckoutResult", "errorMessage :${result.paymentError?.errorMessage}")
//    }
//
//    if (result.isCanceled) {
//      Log.i("handleCheckoutResult", "isCanceled")
//    }
//
//    Log.i("handleCheckoutResult", "${result.transaction}")
//
//    val resourcePath = result.resourcePath
//
//    if (resourcePath != null) {
//      Log.i("resourcePath", "$resourcePath")
//    }
    val item: MutableMap<String, Any?> = HashMap()
    item["isErrored"] = result.isErrored
    val paymentError: MutableMap<String, Any?> = HashMap()
    paymentError["errorCode"] = result.paymentError?.errorCode.toString()
    paymentError["errorInfo"] = result.paymentError?.errorInfo
    paymentError["errorMessage"] = result.paymentError?.errorMessage
    item["paymentError"] = paymentError
    item["isCanceled"] = result.isCanceled
    item["resourcePath"] = result.resourcePath.toString()
    val transaction: MutableMap<String, Any?> = HashMap()
    transaction["transactionType"] = result.transaction?.transactionType.toString()
    val paymentParams: MutableMap<String, Any?> = HashMap()
    paymentParams["checkoutId"] = result.transaction?.paymentParams?.checkoutId
    paymentParams["paymentBrand"] = result.transaction?.paymentParams?.paymentBrand
    paymentParams["shopperResultUrl"] = result.transaction?.paymentParams?.shopperResultUrl
    transaction["paymentParams"] = paymentParams
    transaction["brandSpecificInfo"] = result.transaction?.brandSpecificInfo.toString()
    transaction["redirectUrl"] = result.transaction?.redirectUrl.toString()
    transaction["threeDS2Info"] = result.transaction?.threeDS2Info.toString()
    transaction["threeDS2MethodRedirectUrl"] = result.transaction?.threeDS2MethodRedirectUrl.toString()
    transaction["yooKassaInfo"] = result.transaction?.yooKassaInfo.toString()
    item["transaction"] = transaction
    handleCheckoutResultSink?.success(item)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    binding.addActivityResultListener(this)
    paymentProvider = OppPaymentProvider(activity, providerMode)
    paymentProvider!!.setThreeDSWorkflowListener { activity }

  }

  override fun onDetachedFromActivityForConfigChanges() {
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
    binding.addActivityResultListener(this)
  }

  override fun onDetachedFromActivity() {

  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
    if (requestCode == checkoutRequestCode) {
      // Ensure the result is for the checkout request
      if (data != null) {
        val checkoutResult = parseCheckoutResult(resultCode,data)
        handleCheckoutResult(checkoutResult)
        return true
      }
    }
    return false
  }

  private fun parseCheckoutResult(resultCode: Int,intent: Intent): CheckoutActivityResult {
    val builder = CheckoutActivityResult.Builder()

    val isCanceled = resultCode == 101

    return builder.setTransaction(intent.getParcelableExtra<Parcelable>("com.oppwa.mobile.connect.checkout.dialog.CHECKOUT_RESULT_TRANSACTION") as Transaction?)
      .setPaymentError(intent.getParcelableExtra<Parcelable>("com.oppwa.mobile.connect.checkout.dialog.CHECKOUT_RESULT_ERROR") as PaymentError?)
      .setResourcePath(intent.getStringExtra("com.oppwa.mobile.connect.checkout.dialog.CHECKOUT_RESULT_RESOURCE_PATH"))
      .setCanceled(isCanceled)
      .build()
  }


//  private fun requestCheckoutInfo(checkoutId: String) {
//    try {
//      paymentProvider!!.requestCheckoutInfo(checkoutId, object : ITransactionListener {
//        override fun transactionCompleted(p0: Transaction) {
//          // Handle transaction completion if needed
//        }
//
//        override fun transactionFailed(p0: Transaction, p1: PaymentError) {
//          // Handle transaction failure if needed
//        }
//      })
//    } catch (e: PaymentException) {
//      e.message?.let {
//        val item: MutableMap<String, Any?> = HashMap()
//        item["isErrored"] = true
//        val paymentError: MutableMap<String, Any?> = HashMap()
//        paymentError["errorCode"] = e.error.errorCode.toString()
//        paymentError["errorInfo"] = e.error.errorInfo
//        paymentError["errorMessage"] = e.error.errorMessage
//        item["paymentError"] = paymentError
//        customUIResultSink?.success(item)
//      }
//    }
//  }


  private fun pay(checkoutId: String, cardHolder: String,cardNumber: String,cardExpiryMonth: String,cardExpiryYear: String,cardCVV: String,cardBrand: String,shopperResultUrl: String) {
    try {
      val paymentParams = CardPaymentParams(
        checkoutId,
        cardBrand,
        cardNumber,
        cardHolder,
        cardExpiryMonth,
        "20$cardExpiryYear",
        cardCVV
      )

      paymentParams.shopperResultUrl = shopperResultUrl
      val transaction = Transaction(paymentParams)

      paymentProvider!!.submitTransaction(transaction, object : ITransactionListener {
        override fun transactionCompleted(p0: Transaction) {
          val item: MutableMap<String, Any?> = HashMap()
          item["isErrored"] = false
          val transaction: MutableMap<String, Any?> = HashMap()
          transaction["transactionType"] = p0.transactionType.toString()
          val paymentParams: MutableMap<String, Any?> = HashMap()
          paymentParams["checkoutId"] = p0.paymentParams.checkoutId
          paymentParams["paymentBrand"] = p0.paymentParams.paymentBrand
          paymentParams["shopperResultUrl"] = p0.paymentParams.shopperResultUrl
          transaction["paymentParams"] = paymentParams
          transaction["brandSpecificInfo"] = p0.brandSpecificInfo.toString()
          transaction["redirectUrl"] = p0.redirectUrl.toString()
          transaction["threeDS2Info"] = p0.threeDS2Info.toString()
          transaction["threeDS2MethodRedirectUrl"] = p0.threeDS2MethodRedirectUrl.toString()
          transaction["yooKassaInfo"] = p0.yooKassaInfo.toString()
          item["transaction"] = transaction
          activity.runOnUiThread {
            customUIResultSink?.success(item)
          }
        }

        override fun transactionFailed(p0: Transaction, p1: PaymentError) {
          val item: MutableMap<String, Any?> = HashMap()
          item["isErrored"] = true
          val paymentError: MutableMap<String, Any?> = HashMap()
          paymentError["errorCode"] = p1.errorCode.toString()
          paymentError["errorInfo"] = p1.errorInfo
          paymentError["errorMessage"] = p1.errorMessage
          item["paymentError"] = paymentError
          val transaction: MutableMap<String, Any?> = HashMap()
          transaction["transactionType"] = p0.transactionType.toString()
          val paymentParams: MutableMap<String, Any?> = HashMap()
          paymentParams["checkoutId"] = p0.paymentParams.checkoutId
          paymentParams["paymentBrand"] = p0.paymentParams.paymentBrand
          paymentParams["shopperResultUrl"] = p0.paymentParams.shopperResultUrl
          transaction["paymentParams"] = paymentParams
          transaction["brandSpecificInfo"] = p0.brandSpecificInfo.toString()
          transaction["redirectUrl"] = p0.redirectUrl.toString()
          transaction["threeDS2Info"] = p0.threeDS2Info.toString()
          transaction["threeDS2MethodRedirectUrl"] = p0.threeDS2MethodRedirectUrl.toString()
          transaction["yooKassaInfo"] = p0.yooKassaInfo.toString()
          item["transaction"] = transaction
          activity.runOnUiThread {
            customUIResultSink?.success(item)
          }
        }
      })

      paymentProvider!!.setThreeDSWorkflowListener { activity }

    } catch (e: PaymentException) {
      val item: MutableMap<String, Any?> = HashMap()
      item["isErrored"] = true
      val paymentError: MutableMap<String, Any?> = HashMap()
      paymentError["errorCode"] = e.error.errorCode.toString()
      paymentError["errorInfo"] = e.error.errorInfo
      paymentError["errorMessage"] = e.error.errorMessage
      item["paymentError"] = paymentError
      activity.runOnUiThread {
        customUIResultSink?.success(item)
      }
    }
  }

//  val threeDSConfig = OppThreeDSConfig.Builder()

//  private val threeDSWorkflowListener: ThreeDSWorkflowListener = object : ThreeDSWorkflowListener {
//    override fun onThreeDSChallengeRequired(): Activity {
//      // provide your Activity
//      return activity
//    }
//
//    override fun onThreeDSConfigRequired(): OppThreeDSConfig {
//      // provide your OppThreeDSConfig
//      return threeDSConfig
//    }
//  }

}
